packages/worker/src/dependencies.ts

```diff
@@ -0,0 +1,83 @@
+import ivm from 'isolated-vm';
+import { ApplyMode, Dependencies, Dependency, DependencyFunction, WorkflowInfo } from '@temporalio/workflow';
+
+export { ApplyMode };
+
+/** Return T if T is not a Promise otherwise extracts the type wrapped in T */
+export type StripPromise<T> = T extends Promise<infer I> ? I : T;
+
+type RefParameters<T extends (...args: any) => any> = T extends (...args: infer P) => any
+  ? { [I in keyof P]: ivm.Reference<P[I]> }
+  : never;
+
+export type InjectedDependencyFunctionIvmVariant<F extends DependencyFunction, A extends ApplyMode, R> =
+  | {
+      arguments: 'copy';
+      fn(info: WorkflowInfo, ...args: Parameters<F>): R;
+      applyMode: A;
+      callDuringReplay?: boolean;
+    }
+  | {
+      arguments: 'reference';
+      fn(info: WorkflowInfo, ...args: RefParameters<F>): R;
+      applyMode: A;
+      callDuringReplay?: boolean;
+    };
+
+/**
+ * Takes a {@link DependencyFunction} and turns it into a type safe specification consisting of the function implementation type and call configuration.
+ *
+ * `InjectedDependencyFunction` consists of these attributes:
+ *
+ * - `fn` - type of the implementation function for dependency `F`
+ * - `applyMode` - defines how `fn` is called from the Workflow isolate (@see {@link ApplyMode})
+ * - `callDuringReplay` - whether or not a `fn` will be called during Workflow replay - defaults to `false`
+ * - `arguments` - configure how arguments are transferred between isolates
+ *   - only relevant to `SYNC_*` {@link ApplyMode}s
```

Comment:
Ideally we would only enforce that `arguments` is only specified in sync call modes (and isn't needed at all if not) in the type system, but not a big deal if that really uglies things up

Reply:
Yes, this is only applicable in `SYNC_*` modes, I'll fix the doc.

This type is quite complex so it's hard to understand what it does from reading it.